### PR TITLE
Safe BLS module

### PR DESCRIPTION
**Pull request by blakecduncan**
Blocked by: https://github.com/getwax/wax/pull/52

Adding a Safe module that uses BLS sigs to verify a transaction. I added a hardhat test that submits a userOp to the entrypoint directly.  This way we don't need to run a bundler or start a node to test a module.

I'm wondering if we should reorganize our test directory.  Right now we have three forms of tests.
- **hardhat test** - all you need is `yarn hardhat test` and the tests should work. Hardhat tests are useful for verification types that are hard/not possible in solidity.
- **Integration test** - Need to run a bundler and a node separately for the tests to pass.
- **Forge tests** - just need `forge test`, and tests are written in solidity.

Should rename/organise these differently?  For example, the two ideas I have are:

- `test/hardhat`
- `test/forge`
- `test/integration`

or

- `test/hardhat`
- `test/hardhat/integration` (since the integration tests use hardhat)
- `test/forge`

### CODE REVIEW

File: account-integrations/safe/src/SafeBlsPlugin.sol [hunk 1]
```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity >=0.7.0 <0.9.0;
+pragma abicoder v2;
```
Comment (JohnGuilding) [inline account-integrations/safe/src/SafeBlsPlugin.sol:3]: Seen that this is included in the `SafeECDSAPlugin` too. I believe abicoder was defaulted to v2 as of [solidity 0.8.0](https://docs.soliditylang.org/en/v0.8.21/080-breaking-changes.html#silent-changes-of-the-semantics). I think it makes sense here and in the ECDSA plugin as we're using solidity`>=0.7.0`. Curious if we want to keep supporting versions of solidity less than 0.8.0? Open question don't really have a strong opinion

File: account-integrations/safe/src/SafeBlsPlugin.sol [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    function owner() public pure returns (uint256[4] memory _blsPublicKey) {
+        return _blsPublicKey;
+    }
+
+    function _validateSignature(
+        UserOperation calldata userOp,
+        bytes32 userOpHash
+    ) internal view override returns (uint256 validationData) {
+        require(userOp.signature.length == 64, "VG: Sig bytes length must be 64");
```
Comment (JohnGuilding) [inline account-integrations/safe/src/SafeBlsPlugin.sol:84]: What do you think about custom errors as opposed to require statements with strings? https://soliditylang.org/blog/2021/04/21/custom-errors/

File: account-integrations/safe/src/utils/SafeTypechainTypesHelper.sol
Comment (JohnGuilding) [inline account-integrations/safe/src/utils/SafeTypechainTypesHelper.sol:1]: Now that we are importing 4337 contracts, should we rename this to something more generic? Maybe something like `TypesHelper`?

File: account-integrations/safe/test/hardhat/SafeBlsPlugin.test.ts [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    const userWallet = ethers.Wallet.fromPhrase(MNEMONIC).connect(
+      provider
+    );
+
+    return {
+        provider,
+        userWallet,
+        entryPoint,
+        safe,
+        safeProxyFactory,
```
Comment (JohnGuilding) [inline account-integrations/safe/test/hardhat/SafeBlsPlugin.test.ts:43]: If we're creating a `safeProxyFactory` variable and others before the describe block and then assigning to them, is there a reason to return them in this function as well?

**Review by JohnGuilding — APPROVED** [review]
Looks good - Approving!

Left a few Qs

File: account-integrations/safe/src/SafeBlsPlugin.sol (hunk 1 above)
Comment (blakecduncan) [reply account-integrations/safe/src/SafeBlsPlugin.sol:3]: Sorry just getting around to these comments.  Yeah, I don't have a preference either.  I haven't really done a deep dive to know if it's still necessary to support the older versions

File: account-integrations/safe/src/SafeBlsPlugin.sol (hunk 2 above)
Comment (blakecduncan) [reply account-integrations/safe/src/SafeBlsPlugin.sol:84]: Good call, I'll make the update

**Pull request merged**